### ISSUE TITLE
[question] how to handle FileWrites, IncrementalClean, and Clean

### ISSUE THREAD

**Issue by jonathanpeppers** [microsoft/msbuild#3916]
Xamarin.Android has a few targets that run after `Build`, for example:

* `_GenerateJava` - takes `Foo.dll` -> `Foo.java` in `$(IntermediateOutputPath)`
* `_CompileJava` - takes `Foo.java` -> `Foo.jar` in `$(IntermediateOutputPath)`
* `_CompileDex` - takes `Foo.jar` -> `Foo.dex` in `$(IntermediateOutputPath)`

_NOTE: these are simplified for illustration_

We have another target, `SignAndroidPackage`, that is invoked by IDEs when an APK needs to be produced. It depends on `Build`.

`SignAndroidPackage` - takes `Foo.dll;Foo.dex` -> `Foo.apk`  in `$(OutputPath)`

### Steps to reproduce

Clone this repository: https://github.com/jonathanpeppers/MSBuildIncrementalClean

Directory contents:
```
Foo.cs
Foo.csproj
Xamarin.Android.targets
```

Command line
```
.\build.ps1
```
(script included)

### Expected  behavior

Xamarin.Android should be able to generate intermediate files after `Build` that get added to `FileWrites` and `FileListAbsolute.txt`. A `Clean` should be able to delete these files as well.

### Actual behavior

There are a lot of *weird hacks* in Xamarin.Android's MSBuild targets that get these scenarios mostly working:
- Write to `$(CleanFile)` directly... :man_facepalming:
- Add to `FileWrites` with wildcards
- Add a new target to `$(CleanDependsOn)` that deletes known files our build produces.

We want to clean these up appropriately. Asking for guidance, thanks!

### Environment data
`msbuild /version` output: 

    Microsoft (R) Build Engine version 15.8.169+g1ccb72aefa for .NET Framework

OS info: Windows 10

If applicable, version of the tool that invokes MSBuild (Visual Studio).

**Comment by cdmihai** [comment]
As far as I know, without having looked too much into this, just recording all the file writes in the `FileWrites` item should be enough, as long as the projects get to import `Microsoft.Common.CurrentVersion.Targets` which has a [target](https://github.com/Microsoft/msbuild/blob/8ee86726d01aef0ca2585568268b0533a38115de/src/Tasks/Microsoft.Common.CurrentVersion.targets#L4877-L4950) that dumps the written files in a file which gets read during clean. Clean is recursive, so as long as discoverable projects import the above project file, it should work.

One approach would be to remove all your hacks, implement the above, and understand / fix the issues preventing clean. Some of them could even be bugs in msbuild's common props / targets.

**Comment by jonathanpeppers** [comment]
Let me rephrase. I think I can make the two problems I have more clear.

## Problem 1

If you add a target that needs to run after `Build`, such as:
```
<BuildDependsOn>
  $(BuildDependsOn);
  _SomeTarget;
</BuildDependsOn>
```

If `_SomeTarget` adds to `FileWrites`, the files aren't actually used because `IncrementalClean` runs during `Build`. You can fix this by adding `BeforeTargets="_CleanGetCurrentAndPriorFileWrites"`, but that doesn't _seem_ right (because it's a private target). `BeforeTargets="IncrementalClean"` wasn't enough.

Is there a more appropriate way to run a target *after* `Build`?

Here are binlogs from my linked example above: [problem1.zip](https://github.com/Microsoft/msbuild/files/2567420/problem1.zip)

## Problem 2

If you have a public target that runs in addition to `Build`, let's say named `SomeTarget`. It has `DependsOnTargets="Build"`, and is invoked by the IDE, etc. It has the same issues as no. 1.

It also has the problem if it adds to `FileWrites`, a subsequent `Build` won't add any files generated by `SomeTarget` to `FileWrites`.

What is the appropriate way to handle this problem? Add an additional target that looks for intermediate outputs from `SomeTarget` and adds them to `FileWrites` proactively?

**Comment by jonathanpeppers** [comment]
I came up with a solution for this: https://github.com/jonathanpeppers/MSBuildIncrementalClean#solution

If anyone has input, that would be great!

**Comment by KirillOsenkov** [comment]
Potentially related: https://github.com/dotnet/msbuild/issues/9709
